compiler/quilt/tools/data_transfer.py

```diff
@@ -0,0 +1,282 @@
+    pass
+
+
+def create_s3_session():
+    """
+    Creates a session with automatic retries on 5xx errors.
+    """
+    sess = requests.Session()
+    retries = Retry(total=3,
+                    backoff_factor=.5,
+                    status_forcelist=[500, 502, 503, 504])
+    sess.mount('https://', HTTPAdapter(max_retries=retries))
+    return sess
+
+def download_fragments(store, obj_urls, obj_sizes):
+    assert len(obj_urls) == len(obj_sizes)
+
+    if not obj_urls:
+        print("All fragments are already downloaded!")
+        return True
+
+    obj_queue = sorted(iteritems(obj_urls), reverse=True)
+    total = len(obj_queue)
+    # Some objects might be missing a size; ignore those for now.
+    total_bytes = sum(size or 0 for size in itervalues(obj_sizes))
+
+    downloaded = []
+    lock = Lock()
+
+    print("Downloading %d fragments (%d bytes before compression)..." % (total, total_bytes))
+
+    with tqdm(total=total_bytes, unit='B', unit_scale=True) as progress:
+        def _worker_thread():
+            with create_s3_session() as s3_session:
+                while True:
+                    with lock:
+                        if not obj_queue:
+                            break
+                        obj_hash, url = obj_queue.pop()
+                        original_size = obj_sizes[obj_hash] or 0  # If the size is unknown, just treat it as 0.
+
+                    success = False
+
+                    temp_path_gz = store.temporary_object_path(obj_hash + '.gz')
+                    with open(temp_path_gz, 'ab') as output_file:
+                        for attempt in range(S3_TIMEOUT_RETRIES):
+                            try:
+                                starting_length = output_file.tell()
+                                response = s3_session.get(
+                                    url,
+                                    headers={
+                                        'Range': 'bytes=%d-' % starting_length
+                                    },
+                                    stream=True,
+                                    timeout=(S3_CONNECT_TIMEOUT, S3_READ_TIMEOUT)
+                                )
+
+                                # RANGE_NOT_SATISFIABLE means, we already have the whole file.
+                                if response.status_code == requests.codes.RANGE_NOT_SATISFIABLE:
+                                    with lock:
+                                        progress.update(original_size)
+                                else:
+                                    if not response.ok:
+                                        message = "Download failed for %s:\nURL: %s\nStatus code: %s\nResponse: %r\n" % (
+                                            obj_hash, response.request.url, response.status_code, response.text
+                                        )
+                                        with lock:
+                                            tqdm.write(message)
```

Comment:
Probably for another PR, but maybe it'd be cleaner to factor out these prints in case the function is called from somewhere other than the CLI? That might be hard to do with the threading though.

Reply:
Yeah, this one is going to be really difficult to factor out...